lib/SPIRV/OCLTypeToSPIRV.cpp

```diff
@@ -0,0 +1,303 @@
+//
+//                     The LLVM/SPIRV Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2014 Advanced Micro Devices, Inc. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of Advanced Micro Devices, Inc., nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements adaptation of OCL types for SPIRV.
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "cltytospv"
+
+#include "OCLTypeToSPIRV.h"
+#include "SPIRVInternal.h"
+#include "OCLUtil.h"
+
+#include "llvm/Pass.h"
+#include "llvm/PassSupport.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <set>
+#include <iterator>
+
+using namespace llvm;
+using namespace SPIRV;
+using namespace OCLUtil;
+
+namespace SPIRV {
+
+char OCLTypeToSPIRV::ID = 0;
+
+OCLTypeToSPIRV::OCLTypeToSPIRV()
+  :ModulePass(ID), M(nullptr), Ctx(nullptr), CLVer(0) {
+  initializeOCLTypeToSPIRVPass(*PassRegistry::getPassRegistry());
+}
+
+void
+OCLTypeToSPIRV::getAnalysisUsage(AnalysisUsage& AU) const {
+  AU.setPreservesAll();
+}
+
+bool
+OCLTypeToSPIRV::runOnModule(Module& Module) {
+  DEBUG(dbgs() << "Enter OCLTypeToSPIRV:\n");
+  M = &Module;
+  Ctx = &M->getContext();
+  auto Src = getSPIRVSource(&Module);
+  if (std::get<0>(Src) != spv::SourceLanguageOpenCL_C)
+    return false;
+
+  CLVer = std::get<1>(Src);
+  if (CLVer > kOCLVer::CL20)
```

Comment:
Do you think this check is necessary?
In my opinion check for source language is good enough. If we have another version of C language I would expect to see the same LLVM types for the next version too.